eng/pipelines/templates/stages/code-coverage-upload.yml

```diff
@@ -27,36 +27,74 @@ stages:
           Artifacts: ${{ parameters.DownloadArtifacts }}
 
       - pwsh: |
-          New-Item -ItemType Directory -Force -Path cover
-          New-Item -ItemType Directory -Force -Path cover-int
-          New-Item -ItemType Directory -Force -Path cover-unit
-
-          $unitCoverage = (Get-ChildItem cover-*/unit).FullName -join ","
-          $integrationCoverage = (Get-ChildItem cover-*/int).FullName -join ","
-
-          # Merge unit test coverage across platforms
-          go tool covdata merge -i="$unitCoverage" -o cover-unit
-          if ($LASTEXITCODE -ne 0) { exit $LASTEXITCODE }
-
-          # Merge integration test coverage across platforms
-          go tool covdata merge -i="$integrationCoverage" -o cover-int
-          if ($LASTEXITCODE -ne 0) { exit $LASTEXITCODE }
-
-          # Merge unit and integration code coverage
-          go tool covdata merge -i="cover-unit,cover-int" -o cover
-          if ($LASTEXITCODE -ne 0) { exit $LASTEXITCODE }
-
-          # Convert to text format
-          go tool covdata textfmt -i=cover -o cover.out
-          if ($LASTEXITCODE -ne 0) { exit $LASTEXITCODE }
+          New-Item -ItemType Directory -Force -Path merged-coverage
+          
+          # Find all coverage.out files from different platforms and test types
+          $unitCoverageFiles = @()
+          $integrationCoverageFiles = @()
+          
+          # Collect unit test coverage files from all platforms
+          Get-ChildItem "cover-*/unit/coverage.out" -ErrorAction SilentlyContinue | ForEach-Object {
+              $unitCoverageFiles += $_.FullName
+              Write-Host "Found unit coverage file: $($_.FullName)"
+          }
+          
+          # Collect integration test coverage files from all platforms  
+          Get-ChildItem "cover-*/int/coverage.out" -ErrorAction SilentlyContinue | ForEach-Object {
+              $integrationCoverageFiles += $_.FullName
+              Write-Host "Found integration coverage file: $($_.FullName)"
+          }
+          
+          # Merge unit test coverage files
+          if ($unitCoverageFiles.Count -gt 0) {
+              Write-Host "Merging $($unitCoverageFiles.Count) unit test coverage files..."
+              $unitFilesParam = $unitCoverageFiles -join ","
+              pwsh -File "$(Build.SourcesDirectory)/eng/scripts/Merge-GoCoverageProfiles.ps1" -InputFiles $unitFilesParam -OutputFile "merged-coverage/unit-coverage.out" -NormalizePaths
+              if ($LASTEXITCODE -ne 0) { 
+                  Write-Error "Failed to merge unit test coverage files"
```

Comment:
What was the motivating factor to remove the `go tool covdata` invocations? From a quick look, it seems that [covdata](https://pkg.go.dev/cmd/covdata) is still supported today.

Reply:
My mistake. As I was reading and understanding why we use `go tool covdata`, I though that in go 1.21, we could have migrated to use coverageprofile instead. But I am reading again and it looks like that's the legacy and gocoverdir is better with binary files and merging with covdata.

I'll revert those changes.